### ISSUE TITLE
1.7.1rc1 seems to keep re-scheduling successful tasks

### ISSUE THREAD

**Issue by garthcn** [apache/airflow#1365]
Please note that this is only happening to us on 1.7.1rc1, which is not an official release. 
Reverting to 1.7.0 seems to have solved the issue. 
### Environment
- Version of Airflow: 1.7.1rc1
- Airflow components and configuration, if applicable: CeleryExecutor
### Description of Issue

We have a Pool (size 4) for our task instances. Airflow is supposed to add task instances to the queue, pick at most 4 from the queue, finish them, and get more task instances from the queue. This is working properly before 1.7.1rc1. 

However, after updating to 1.7.1rc1, Airflow adds tasks instances to the queue, takes 4 from the queue, finishes them, but stops taking more from the queue. Instead, it would re-add successful task instances repeatedly, try to run them, and exit without doing anything (because it would find out that it's already successful). In each task instance's log file, there are many lines like:

```
[2016-04-12 19:15:45,242] {models.py:1042} INFO - Task <TaskInstance: dag.dag_node 2016-04-12 05:00:00 [success]> previously succeeded on 2016-04-12 19:15:15
```

It keeps running each already finished task instances, and keeps printing out lines like this. 
### Reproduce

I'm not sure if this is reliably reproducible, but our setup is:
1. Create a Pool of size 4 (or any number smaller than ready-to-run task instances)
2. Start Scheduler so that it adds more than 4 to the queue
3. Wait until Airflow finishes running the first batch of 4 task instances
4. It will start re-adding and re-running successful task instances, and never take un-run tasks from the queue. You can observe it in `airflow flower` UI.

**Comment by r39132** [comment]
Thanks for reporting this @garthcn ( I will try to reproduce this shortly ) cc/ @mistercrunch @criccomini @jlowin

**Comment by criccomini** [comment]
@r39132 I think this might be the same thing that I'm seeing. When I grep for previously succeeded on, I find a ton of entries after upgrading to 1.7.1.

**Comment by criccomini** [comment]
Note: I was using local executor, so problem seems to happen both on celery and local.

**Comment by jlowin** [comment]
I might have found the problem. I was making many changes regarding the scheduler's behavior when this release candidate was cut and it looks like one of my PRs made it in and one just missed -- and the one that missed included a bugfix for this situation. Specifically, this one line (no. 823) https://github.com/airbnb/airflow/pull/1290/files?diff=unified#diff-54a57ccc2c8e73d12c812798bf79ccb2R823

Without this line, the Scheduler's set of queued TI's doesn't clear even when they succeed, and the Scheduler tries to put them through again and again.

BUT this morning @abridgett raised a situation that may have been caused by the new logic when used in conjunction with `airflow scheduler --num-runs`, and I just submitted #1378 which undoes the changes in question and completely restores 1.6.2 behavior for queued tasks. 

Long story short: don't use 1.7.1rc :disappointed:

**Comment by jlowin** [comment]
by the way @r39132 this was my test dag. @garthcn Does it exhibit the behavior for you too? I actually did NOT see the "previously succeeded" message in the logs for some reason, but I could see that it was trying to run the tasks.

``` python
import airflow
from airflow import DAG
from airflow.models import Pool
from airflow.operators import BashOperator
import datetime

session = airflow.settings.Session()

if not (
        session.query(Pool)
            .filter(Pool.pool == 'pool')
            .first()):
    session.add(Pool(pool='pool', slots=4))


default_args = dict(owner='airflow', start_date=datetime.datetime(2016, 4, 12))
dag = DAG('dag', default_args=default_args)

for i in range(3):
    BashOperator(
        task_id='op_{}'.format(i),
        bash_command="sleep 1",
        dag=dag,
        pool='pool')
```

**Comment by bolkedebruin** [comment]
@garthcn ping. It would be nice to have this fixed before a new release

**Comment by garthcn** [comment]
I'm not using `airflow scheduler --num-runs`. And I see tons of `previously succeeded` in each task instance. I can confirm that in my case they are being scheduled repeatedly.

**Comment by jlowin** [comment]
Ok, in that case #1290 (which unfortunately just missed the rc cutoff) will fix your specific issue and #1378 will address it further for people using `--num-runs`.

**Comment by bolkedebruin** [comment]
Please reopen if issue continues to persist

**Comment by r39132** [comment]
I took @jlowin's code above and started running.

Git log looks like this:
```
commit 7da6a94ebc77b6b763485c5f9072a8ebce5af8d7
Merge: 8b63891 a36861a
Author: bolkedebruin <bolkedebruin@users.noreply.github.com>
Date:   Sun Apr 17 18:51:35 2016 +0200

    Merge pull request #1376 from bolkedebruin/multiprocessing_scheduler
    
    Use multiprocessing for the scheduler

commit a36861a5f7f296656302942db83dd65a66dd5949
Author: Bolke de Bruin <bolke@xs4all.nl>
Date:   Sat Apr 16 13:03:50 2016 +0200

    Add multiprocessing support to the scheduler
    
    As the amount of dags grows and the ability to create dags programmatically
    is more often used, more and more time is spend in the scheduler which lowers
    throughput.
    
    This patch adds the ability to use multiple threads to the scheduler. The
    amount of threads can be specified by "max_threads" in the scheduler section
    of the configuration. The amount of threads will, however, not exceed the
    amount of cores.
    
    In case of using sqlite the max_threads will be set to 1 as sqlite does not
    support multiple db connections.

commit f67cecb3ff3ccc9fd0725d7d6def30040779cd07
```

I'm seeing the following error in the scheduler window:
`[2016-04-17 18:03:53,635] {jobs.py:726} ERROR - Failed at reloading the dagbag. unsupported operand type(s) for %: 'NoneType' and 'int'` 

Though the scheduler is running, I'm seeing no progress beyond the first few attempts and things appear stalled. I suspect it's related to the scheduler error above. 

<img width="905" alt="screenshot 2016-04-17 17 58 54" src="https://cloud.githubusercontent.com/assets/581734/14591489/18543b2c-04c7-11e6-9cb1-e72a7dafa6d6.png">
<img width="885" alt="screenshot 2016-04-17 18 00 29" src="https://cloud.githubusercontent.com/assets/581734/14591488/1853b6c0-04c7-11e6-93b5-2c7c2fa496fa.png">

**Comment by bolkedebruin** [comment]
@r39132 @garthcn please confirm if this issue persists with rc3 _and_ #1378 applied. Thanks.
